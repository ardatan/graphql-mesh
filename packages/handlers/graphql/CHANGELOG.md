# @graphql-mesh/graphql

## 0.14.1

### Patch Changes

- Updated dependencies [191a663a]
  - @graphql-mesh/types@0.35.1

## 0.14.0

### Minor Changes

- b9ca0c30: Make Transforms and Handlers base-dir aware

### Patch Changes

- Updated dependencies [b9ca0c30]
  - @graphql-mesh/types@0.35.0
  - @graphql-mesh/utils@0.9.0

## 0.13.19

### Patch Changes

- Updated dependencies [ec89a923]
  - @graphql-mesh/utils@0.8.8

## 0.13.18

### Patch Changes

- Updated dependencies [55327fd6]
  - @graphql-mesh/types@0.34.1

## 0.13.17

### Patch Changes

- Updated dependencies [76051dd7]
  - @graphql-mesh/types@0.34.0

## 0.13.16

### Patch Changes

- Updated dependencies [646d6bdb]
  - @graphql-mesh/types@0.33.0

## 0.13.15

### Patch Changes

- Updated dependencies [68d6b117]
  - @graphql-mesh/types@0.32.0

## 0.13.14

### Patch Changes

- Updated dependencies [212f2d66]
  - @graphql-mesh/types@0.31.1

## 0.13.13

### Patch Changes

- Updated dependencies [77327988]
  - @graphql-mesh/types@0.31.0

## 0.13.12

### Patch Changes

- Updated dependencies [48f38a4a]
  - @graphql-mesh/types@0.30.1

## 0.13.11

### Patch Changes

- Updated dependencies [938cca26]
  - @graphql-mesh/types@0.30.0

## 0.13.10

### Patch Changes

- Updated dependencies [8ef29de1]
  - @graphql-mesh/types@0.29.4

## 0.13.9

### Patch Changes

- Updated dependencies [a02d86c3]
- Updated dependencies [a02d86c3]
- Updated dependencies [a02d86c3]
  - @graphql-mesh/types@0.29.3

## 0.13.8

### Patch Changes

- Updated dependencies [69d2198d]
  - @graphql-mesh/utils@0.8.7

## 0.13.7

### Patch Changes

- Updated dependencies [8e8848e1]
  - @graphql-mesh/types@0.29.2

## 0.13.6

### Patch Changes

- Updated dependencies [7e970f09]
  - @graphql-mesh/utils@0.8.6

## 0.13.5

### Patch Changes

- Updated dependencies [e8994875]
  - @graphql-mesh/types@0.29.1

## 0.13.4

### Patch Changes

- Updated dependencies [8d345721]
  - @graphql-mesh/utils@0.8.5

## 0.13.3

### Patch Changes

- Updated dependencies [c767df01]
- Updated dependencies [183cfa96]
- Updated dependencies [b3d7ecbf]
  - @graphql-mesh/types@0.29.0
  - @graphql-mesh/utils@0.8.4

## 0.13.2

### Patch Changes

- Updated dependencies [a22fc6f3]
  - @graphql-mesh/types@0.28.0

## 0.13.1

### Patch Changes

- Updated dependencies [c1de3e43]
  - @graphql-mesh/types@0.27.0

## 0.13.0

### Minor Changes

- 75f6dff9: feat(graphql): ability to disable batch execution

### Patch Changes

- Updated dependencies [75f6dff9]
- Updated dependencies [c4f207a7]
  - @graphql-mesh/types@0.26.0

## 0.12.1

### Patch Changes

- d4ed4e89: fix(graphql): schemaHeaders bug

## 0.12.0

### Minor Changes

- 0df817d0: feat(graphql): support exported schemaHeaders

### Patch Changes

- Updated dependencies [0df817d0]
  - @graphql-mesh/types@0.25.0

## 0.11.3

### Patch Changes

- Updated dependencies [08c2966e]
  - @graphql-mesh/utils@0.8.3

## 0.11.2

### Patch Changes

- Updated dependencies [b6262481]
  - @graphql-mesh/types@0.24.0

## 0.11.1

### Patch Changes

- Updated dependencies [e5b38574]
  - @graphql-mesh/types@0.23.3

## 0.11.0

### Minor Changes

- 7140ef09: feat(graphql): support custom variables in endpoint url

## 0.10.5

### Patch Changes

- Updated dependencies [c85a54eb]
  - @graphql-mesh/utils@0.8.2

## 0.10.4

### Patch Changes

- Updated dependencies [c614e796]
  - @graphql-mesh/types@0.23.2

## 0.10.3

### Patch Changes

- Updated dependencies [59d77fb8]
  - @graphql-mesh/types@0.23.1

## 0.10.2

### Patch Changes

- 44e14ca5: fix(graphql): handle null variables in multipart requests

## 0.10.1

### Patch Changes

- Updated dependencies [e5cd44f5]
  - @graphql-mesh/types@0.23.0

## 0.10.0

### Minor Changes

- 2fd59a83: feat(graphql): add useSSEForSubscription option to use SSE for Subscriptions instead of WebSocket

### Patch Changes

- Updated dependencies [2fd59a83]
  - @graphql-mesh/types@0.22.0

## 0.9.2

### Patch Changes

- 86f359cb: fix(graphql): fix handling urls

## 0.9.1

### Patch Changes

- Updated dependencies [c064e3a8]
  - @graphql-mesh/types@0.21.1
  - @graphql-mesh/utils@0.8.1

## 0.9.0

### Minor Changes

- 03f41cd0: feat(graphql): support exported GraphQLSchema from code files

### Patch Changes

- 1f0b2f1f: fix(graphql): fix multipart issues
- Updated dependencies [03f41cd0]
  - @graphql-mesh/types@0.21.0

## 0.8.1

### Patch Changes

- 1e7fd602: feat(graphql): add `multipart` option to support file uploads
- Updated dependencies [1e7fd602]
  - @graphql-mesh/types@0.20.1

## 0.8.0

### Minor Changes

- 2d14fcc3: bump graphql-tools to v7
- 2d14fcc3: feat(graphql): use graphql-ws instead of subscriptions-transport-ws

### Patch Changes

- Updated dependencies [2d14fcc3]
- Updated dependencies [2d14fcc3]
  - @graphql-mesh/types@0.20.0
  - @graphql-mesh/utils@0.8.0

## 0.7.10

### Patch Changes

- Updated dependencies [c1b073de]
  - @graphql-mesh/types@0.19.0
  - @graphql-mesh/utils@0.7.0

## 0.7.9

### Patch Changes

- Updated dependencies [5628fb14]
  - @graphql-mesh/types@0.18.0

## 0.7.8

### Patch Changes

- Updated dependencies [0560e806]
  - @graphql-mesh/types@0.17.1

## 0.7.7

### Patch Changes

- Updated dependencies [c26c8c56]
  - @graphql-mesh/types@0.17.0

## 0.7.6

### Patch Changes

- Updated dependencies [3770af72]
  - @graphql-mesh/types@0.16.1

## 0.7.5

### Patch Changes

- Updated dependencies [3ee10180]
  - @graphql-mesh/types@0.16.0

## 0.7.4

### Patch Changes

- Updated dependencies [0f17c58c]
  - @graphql-mesh/types@0.15.0

## 0.7.3

### Patch Changes

- Updated dependencies [937c87d2]
  - @graphql-mesh/types@0.14.1

## 0.7.2

### Patch Changes

- Updated dependencies [1e0445ee]
  - @graphql-mesh/types@0.14.0

## 0.7.1

### Patch Changes

- Updated dependencies [b50a68e3]
  - @graphql-mesh/types@0.13.0

## 0.7.0

### Minor Changes

- e2b34219: improve introspection caching

### Patch Changes

- Updated dependencies [e2b34219]
- Updated dependencies [9a7a55c4]
  - @graphql-mesh/types@0.12.0
  - @graphql-mesh/utils@0.6.0

## 0.6.4

### Patch Changes

- a3b42cfd: fix(runtime): handle transforms correctly for single source
- Updated dependencies [2dedda3c]
- Updated dependencies [a3b42cfd]
  - @graphql-mesh/types@0.11.3
  - @graphql-mesh/utils@0.5.4

## 0.6.3

### Patch Changes

- Updated dependencies [6d624576]
  - @graphql-mesh/types@0.11.2

## 0.6.2

### Patch Changes

- Updated dependencies [405cec23]
  - @graphql-mesh/types@0.11.1

## 0.6.1

### Patch Changes

- Updated dependencies [48d89de2]
  - @graphql-mesh/types@0.11.0

## 0.6.0

### Minor Changes

- 79adf4b6: feat(config): support functions as config property

### Patch Changes

- Updated dependencies [79adf4b6]
- Updated dependencies [79adf4b6]
  - @graphql-mesh/utils@0.5.3
  - @graphql-mesh/types@0.10.0

## 0.5.5

### Patch Changes

- Updated dependencies [2d5cc25b]
  - @graphql-mesh/types@0.9.2

## 0.5.4

### Patch Changes

- Updated dependencies [93ad5255]
  - @graphql-mesh/types@0.9.1

## 0.5.3

### Patch Changes

- Updated dependencies [9900d2fa]
- Updated dependencies [9900d2fa]
  - @graphql-mesh/utils@0.5.2

## 0.5.2

### Patch Changes

- Updated dependencies [c8d9695e]
- Updated dependencies [8f53be10]
  - @graphql-mesh/types@0.9.0
  - @graphql-mesh/utils@0.5.1

## 0.5.1

### Patch Changes

- Updated dependencies [d2e56567]
  - @graphql-mesh/types@0.8.1

## 0.5.0

### Minor Changes

- a789c312: feat(stitching): use batch execution

### Patch Changes

- Updated dependencies [a789c312]
  - @graphql-mesh/types@0.8.0
  - @graphql-mesh/utils@0.5.0

## 0.4.0

### Minor Changes

- 718e7a16: fix(runtime): fix browser support

### Patch Changes

- Updated dependencies [718e7a16]
  - @graphql-mesh/types@0.7.0
  - @graphql-mesh/utils@0.4.0

## 0.3.2

### Patch Changes

- Updated dependencies [5067ac73]
- Updated dependencies [a76d74bb]
  - @graphql-mesh/types@0.6.0

## 0.3.1

### Patch Changes

- Updated dependencies [dde7878b]
  - @graphql-mesh/types@0.5.1

## 0.3.0

### Minor Changes

- 705c4626: introduce an independent config package

### Patch Changes

- Updated dependencies [705c4626]
  - @graphql-mesh/types@0.5.0
  - @graphql-mesh/utils@0.3.0

## 0.2.19

### Patch Changes

- Updated dependencies [854dc550]
- Updated dependencies [6f21094b]
  - @graphql-mesh/types@0.4.0

## 0.2.18

### Patch Changes

- Updated dependencies [3c131332]
  - @graphql-mesh/types@0.3.1

## 0.2.17

### Patch Changes

- Updated dependencies [ccede377]
  - @graphql-mesh/types@0.3.0

## 0.2.16

### Patch Changes

- 1d32f9bf: Fix handling GraphQL variables with useGETForQueries
