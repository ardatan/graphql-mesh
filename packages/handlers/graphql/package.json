{
  "name": "@graphql-mesh/graphql",
  "version": "0.14.1",
  "sideEffects": false,
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.js",
  "typings": "dist/index.d.ts",
  "typescript": {
    "definition": "dist/index.d.ts"
  },
  "license": "MIT",
  "peerDependencies": {
    "graphql": "*"
  },
  "dependencies": {
    "@graphql-mesh/types": "0.35.1",
    "@graphql-mesh/utils": "0.9.0",
    "fetchache": "0.0.4",
    "@graphql-tools/url-loader": "6.8.2",
    "@graphql-tools/wrap": "7.0.5"
  },
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  }
}
