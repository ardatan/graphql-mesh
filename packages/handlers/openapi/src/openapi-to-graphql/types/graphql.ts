// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
