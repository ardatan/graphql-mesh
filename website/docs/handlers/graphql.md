---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

This handler allows you to load remote GraphQL schemas and use it with schema-stitching, based on `graphql-tools-fork`.

To get started, install the handler library from NPM:

```
$ yarn add @graphql-mesh/graphql
```

Now, you can use it directly in your Mesh config file:

```yml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

## Config API Reference

{@import ../generated-markdown/GraphQLHandler.generated.md}
