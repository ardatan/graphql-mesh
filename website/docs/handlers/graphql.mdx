---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageInstall packages="@graphql-mesh/graphql" />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

> You can check out our example that uses schema stitching with a PostgreSQL data source.
> [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

<p>&nbsp;</p>

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

<p>&nbsp;</p>

### From Environmental Variable

`MY_API_TOKEN` is the name of the environmental variable you have the value.

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Config API Reference

{@import ../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md}
or
{@import ../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md}
