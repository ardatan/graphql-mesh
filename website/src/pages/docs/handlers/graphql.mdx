import { PackageCmd, Callout } from '@theguild/components'

# GraphQL

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageCmd packages={['@graphql-mesh/graphql']} />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

<Callout>
  You can check out our example that uses schema stitching with a PostgreSQL data source.
  [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

</Callout>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          # Use "{context.headers['x-my-api-token']}" if you want just the value of the header
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

### From Environment Variables

`MY_API_TOKEN` is the name of the environment variable you have the value.

```yaml {7}
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

## Local Schemas

We recommend providing local schema by using the `additionalTypeDefs` and `additionalResolvers` configuration options.

However, it is also possible to use a local GraphQL Schema instance as a GraphQL Mesh source, as showcased below:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        schema: ./my-local-schema.ts
```

```ts filename="my-local-schema.ts"
import { makeExecutableSchema } from '@graphql-tools/schema'
export default makeExecutableSchema({
  typeDefs: /* GraphQL */ `
    type Query {
      foo: String
    }
  `,
  resolvers: {
    Query: {
      foo: () => 'FOO'
    }
  }
})
```

## Fetch Strategies and Multiple HTTP endpoints for the same source

If you want to have an advanced fetch strategy for the GraphQL source such as retrying twice or timeout in 30 seconds etc.
Also, you can have different HTTP endpoints for a single source, and you can configure Mesh to get a better execution flow.

For example, you can make a request to both endpoints and return the fastest response with `race` strategy.

All `fetch` strategies can be combined to create the ultimate execution flow:

<details>
 <summary>`retry`</summary>

The `retry` mechanism allow you to specify the retry attempts for a single GraphQL endpoint/source.

The retry flow will execute in both conditions: a network error, or due to a runtime error.

```yaml {6}
sources:
  - name: uniswapv2
    handler:
      graphql:
        endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
        retry: 2 # specify here, if you have an unstable/error prone indexer
```

</details>

<details>
 <summary>`timeout`</summary>

The `timeout` mechanism allow you to specify the `timeout` for a given GraphQL endpoint.

```yaml {6}
sources:
  - name: uniswapv2
    handler:
      graphql:
        endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
        timeout: 5000 # 5 seconds
```

</details>

<details>
 <summary>`fallback`</summary>

The `fallback` mechanism allow you to specify use more than one GraphQL endpoint, for the same source.

This is helpful if you have a fallback endpoint for the same GraphQL API.

```yaml {5}
sources:
  - name: uniswapv2
    handler:
      graphql:
        strategy: fallback
        sources:
          - endpoint: https://bad-uniswap-v2-api.com
            retry: 2
            timeout: 5000
          - endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
```

</details>

<details>
 <summary>`race`</summary>

The `race` mechanism allow you to specify use more than one GraphQL endpoint, for the same source, and race on every execution.

If you have different places that service is deployed, this is useful to get the fastest response by racing them.

```yaml {5}
sources:
  - name: uniswapv2
    handler:
      graphql:
        strategy: race
        sources:
          - endpoint: https://bad-uniswap-v2-api.com
          - endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
```

</details>

## Config API Reference

import API1 from '../../../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md'
import API2 from '../../../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md'

<API1 />
or
<API2 />
