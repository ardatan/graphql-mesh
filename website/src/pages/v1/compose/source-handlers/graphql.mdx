---
description: GraphQL Mesh can load remote GraphQL schemas, fetch from a CDN or registry
---

import { Callout } from '@theguild/components'

# GraphQL

<Callout>
  This page is currently under construction and expected to change. Please feel free to reach out to
  us directly in case you are having any troubles.
</Callout>

This handler allows you to load remote GraphQL schemas and compose them into a supergraph.

## How to use?

If you want to load a remote GraphQL subgraph by introspecting the endpoint, you can only provide
`endpoint` in the configuration like below;

```ts filename="mesh-config.ts"
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Users', {
        endpoint: 'http://localhost:4001/users'
      })
    }
  ]
})
```

### Load the schema via URL or local file (Hive CDN etc)

If you want to load a remote GraphQL subgraph from a CDN, you can provide `source` in the
configuration like below;

In this case, Mesh will take the schema from `source`, but execute the operations against the
`endpoint`.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('GitHub', {
        endpoint: 'https://api.github.com/graphql',
        // You can provide your SDL or introspection seperately
        source: 'https://docs.github.com/public/schema.docs.graphql',

        operationHeaders: {
          // This forwards the header from the incoming request to the remote server
          authorization: 'Bearer {context.headers["x-my-api-token"]}'
        }
      })
    }
  ]
})
```

## Headers

[Read about configuration and examples](/v1/compose/source-handlers#setting-headers)

## Fetch Strategies and Multiple HTTP endpoints for the same source

If you want to have an advanced fetch strategy for the GraphQL source such as retrying twice or
timeout in 30 seconds etc. Also, you can have different HTTP endpoints for a single source, and you
can configure Mesh to get a better execution flow.

For example, you can make a request to both endpoints and return the fastest response with `race`
strategy.

All `fetch` strategies can be combined to create the ultimate execution flow:

<details>
 <summary>`retry`</summary>

The `retry` mechanism allow you to specify the retry attempts for a single GraphQL endpoint/source.

The retry flow will execute in both conditions: a network error, or due to a runtime error.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2',
        // Specify here, if you have an unstable/error prone indexer
        retry: 2
      })
    }
  ]
})
```

</details>

<details>
 <summary>`timeout`</summary>

The `timeout` mechanism allow you to specify the `timeout` for a given GraphQL endpoint.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2',
        // 5 Seconds
        timeout: 5_000
      })
    }
  ]
})
```

</details>

<details>
 <summary>`fallback`</summary>

The `fallback` mechanism allow you to specify use more than one GraphQL endpoint, for the same
source.

This is helpful if you have a fallback endpoint for the same GraphQL API.

```ts filename="mesh-config.ts" {7-17}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        strategy: 'fallback',
        sources: [
          {
            endpoint: 'https://bad-uniswap-v2-api.com',
            retry: 2,
            timeout: 5_000
          },
          {
            endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2'
          }
        ]
      })
    }
  ]
})
```

</details>

<details>
 <summary>`race`</summary>

The `race` mechanism allow you to specify use more than one GraphQL endpoint, for the same source,
and race on every execution.

If you have different places that service is deployed, this is useful to get the fastest response by
racing them.

```ts filename="mesh-config.ts" {7-15}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        strategy: 'race',
        sources: [
          {
            endpoint: 'https://bad-uniswap-v2-api.com'
          },
          {
            endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2'
          }
        ]
      })
    }
  ]
})
```

</details>
