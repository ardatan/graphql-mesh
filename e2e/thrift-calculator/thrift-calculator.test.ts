import { parse } from 'graphql';
import { createTenv } from '@e2e/tenv';
import { getExecutorForFusiongraph } from '@graphql-mesh/fusion-runtime';

const { compose, service } = createTenv(__dirname);

it('should compose the appropriate schema', async () => {
  const { result } = await compose({
    services: [await service('calculator')],
    maskServicePorts: true,
  });
  expect(result).toMatchSnapshot();
});

it.concurrent.each([
  {
    name: 'Add',
    document: parse(/* GraphQL */ `
      query Add {
        add(request: { left: 2, right: 3 })
      }
    `),
  },
])('should execute $name', async ({ document }) => {
  const { result } = await compose({
    services: [await service('calculator')],
  });

  const { fusiongraphExecutor } = getExecutorForFusiongraph({ fusiongraph: result });

  await expect(fusiongraphExecutor({ document })).resolves.toMatchSnapshot();
});
