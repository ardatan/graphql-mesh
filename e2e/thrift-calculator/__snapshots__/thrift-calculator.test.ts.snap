// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compose the appropriate schema 1`] = `
"
schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION)
  
  
  
  
  
  @link(
  url: "https://the-guild.dev/graphql/mesh/spec/v1.0"
  import: ["@fieldTypeMap", "@extraSchemaDefinitionDirective"]
)
{
  query: Query
  
  
}


  directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

  directive @join__field(
    graph: join__Graph
    requires: join__FieldSet
    provides: join__FieldSet
    type: String
    external: Boolean
    override: String
    usedOverridden: Boolean
  ) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

  directive @join__graph(name: String!, url: String!) on ENUM_VALUE

  directive @join__implements(
    graph: join__Graph!
    interface: String!
  ) repeatable on OBJECT | INTERFACE

  directive @join__type(
    graph: join__Graph!
    key: join__FieldSet
    extension: Boolean! = false
    resolvable: Boolean! = true
    isInterfaceObject: Boolean! = false
  ) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

  directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

  scalar join__FieldSet


  directive @link(
    url: String
    as: String
    for: link__Purpose
    import: [link__Import]
  ) repeatable on SCHEMA

  scalar link__Import

  enum link__Purpose {
    """
    \`SECURITY\` features provide metadata necessary to securely resolve fields.
    """
    SECURITY

    """
    \`EXECUTION\` features provide metadata necessary for operation execution.
    """
    EXECUTION
  }







enum join__Graph {
  CALCULATOR @join__graph(name: "calculator", url: "http://0.0.0.0:<calculator_port>/thrift") 
}

directive @fieldTypeMap(subgraph: String, fieldTypeMap: FieldTypeMap)  repeatable on FIELD_DEFINITION

directive @extraSchemaDefinitionDirective(directives: _DirectiveExtensions)  repeatable on OBJECT

scalar FieldTypeMap @join__type(graph: CALCULATOR) 

scalar _DirectiveExtensions @join__type(graph: CALCULATOR) 

type Query @extraSchemaDefinitionDirective(
  directives: {transport: [{subgraph: "calculator", kind: "thrift", location: "http://0.0.0.0:<calculator_port>/thrift", headers: {}, options: {clientAnnotations: {serviceName: "Calculator", annotations: {}, methodNames: ["add", "subtract"], methodAnnotations: {add: {annotations: {}, fieldAnnotations: {}}, subtract: {annotations: {}, fieldAnnotations: {}}}, methodParameters: {add: 2, subtract: 3}}, topTypeMap: {AddRequest: {id: 0, name: "AddRequest", type: 12, fields: {left: {type: 8, id: 1}, right: {type: 8, id: 2}}}}}}]}
) @join__type(graph: CALCULATOR)  {
  add(request: AddRequestInput) : Int @fieldTypeMap(
    subgraph: "calculator"
    fieldTypeMap: {request: {name: "AddRequest", type: "ref", id: 1}}
  )
  subtract(left: Int, right: Int) : Int @fieldTypeMap(
    subgraph: "calculator"
    fieldTypeMap: {left: {type: 8, id: 1}, right: {type: 8, id: 2}}
  )
}

input AddRequestInput @join__type(graph: CALCULATOR)  {
  left: Int!
  right: Int!
}
    
"
`;

exports[`should execute Add 1`] = `
{
  "data": {
    "add": 5,
  },
}
`;
